### PR TITLE
Adds support for templates index pattern

### PR DESCRIPTION
**Pull request by jalvz**
See motivation in https://github.com/elastic/package-spec/issues/64#issuecomment-755342698

It will require Kibana support, basically if this new property is `true`, the index pattern generated for the templates should be `<type>-<dataset>.*-*`
I will open a ticket in Kibana if this is accepted.

### CODE REVIEW

**Comment by ruflin** [comment]
@jalvz Can you share a bit more details on how this will be used?

**Comment by jalvz** [comment]
@ruflin you have been too fast :) 
Backlink in description now

File: versions/1/data_stream/manifest.spec.yml
Comment (mtojek) [inline versions/1/data_stream/manifest.spec.yml:170]: nit: I prefer to keep a human readable description (without underscore).

File: versions/1/data_stream/manifest.spec.yml
Comment (mtojek) [inline versions/1/data_stream/manifest.spec.yml:169]: Is it possible to define a regular expression for the field value?

**Review by mtojek — COMMENTED** [review]
Please modify the "good" test package, so we can be sure that the change works as intended.

**Review by ruflin — CHANGES_REQUESTED** [review]
I'm currently blocking this PR as we should first complete the discussion in https://github.com/elastic/package-spec/issues/64#issuecomment-755342698

I like the simplicity of this change but I worry it has many side effects. On the Stack side we loose the knowledge if the templates / mappings / index patterns for a data stream are all aligned or if a dev might have just introduced an invalide index pattern.

**Comment by jalvz** [comment]
Updated as per https://github.com/elastic/package-spec/issues/64#issuecomment-757925161

**Comment by ruflin** [comment]
@skh @jfsiii @jen-huang Could you chime in from the Kibana side if this is feasible?
@mtojek @ycombinator Not happy with the name yet (see conversation in the linked issue), so if you have ideas for better names ...
@jalvz LGTM but lets wait on the feedback from the Kibana side to not merge this too soon.

File: versions/1/data_stream/manifest.spec.yml
Comment (simitt) [inline versions/1/data_stream/manifest.spec.yml:114]: How about `dataset_dynamic` or `dataset_is_pattern` as this only changes whether or not the `dataset` part is a string or a pattern)? 
Releated to the description - we could then change it to something like _indicates whether the dataset name can contain a dynamic part only determined during runtime_.

**Comment by jalvz** [comment]
Filed Kibana issue: https://github.com/elastic/kibana/issues/88307

File: versions/1/data_stream/manifest.spec.yml
Comment (skh) [inline versions/1/data_stream/manifest.spec.yml:115]: Is this only about the field `index_patterns` in an ES index template (this is what I guessed from the discussions), or also about Kibana index patterns (this is how this line reads without context)?

Comment (ruflin) [reply versions/1/data_stream/manifest.spec.yml:115]: It is only about the ES index patterns. It should not affect the Kibana ones as these are based on the types and include all datasets anyways.

Comment (skh) [reply versions/1/data_stream/manifest.spec.yml:115]: Then this description should be changed accordingly.

**Comment by jalvz** [comment]
Hey all, are there any more objections to this?

**Review by simitt — COMMENTED** [review]
LGTM

**Review by mtojek — APPROVED** [review]
LGTM, but let's leave the final decision to @ruflin .

EDIT:

please rebase the branch against master and update the changelog.

**Review by ruflin — APPROVED** [review]
LGTM

@mtojek @ycombinator Side discussion: Should this go into https://github.com/elastic/package-spec/blob/master/versions/1/changelog.yml or where is the right place for this?

**Comment by mtojek** [comment]
> @mtojek @ycombinator Side discussion: Should this go into https://github.com/elastic/package-spec/blob/master/versions/1/changelog.yml or where is the right place for this?

Yes, this is the right place.

**Pull request merged**
